### PR TITLE
Endpoints Tab | 01 | Add actions to request SDK Service Endpoints

### PR DESCRIPTION
**Pull request by bstavroulakis**
These are the new actions to receive the SDK Service Endpoints

Closes DCOS-18144

**Checklist**
- [ ] Did you add a JIRA issue in a commit message or as part of the branch name?
- [ ] Did you add new unit tests?
- [ ] Did you add new integration tests?
- [ ] If this is a regression, did you write a test to catch this in the future?

### CODE REVIEW

File: plugins/services/src/js/events/SDKEndpointActions.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  REQUEST_SDK_ENDPOINTS_ERROR,
+  REQUEST_SDK_ENDPOINT_SUCCESS,
+  REQUEST_SDK_ENDPOINT_ERROR,
+  REQUEST_SDK_ENDPOINTS_LOADING
+} from "../constants/ActionTypes";
+
+const SDKEndpointsActions = {
+  fetchEndpoints(serviceId) {
+    AppDispatcher.handleServerAction({
+      type: REQUEST_SDK_ENDPOINTS_LOADING,
```
Comment (MatApple) [inline plugins/services/src/js/events/SDKEndpointActions.js]: I don't think we should introduce this pattern unless there is a strong argument for introducing it. Please follow our existing pattern in react components for identifying loading state or discuss in advance the need for such an introduction, before implementing it :)

Comment (bstavroulakis) [reply plugins/services/src/js/events/SDKEndpointActions.js]: Removed

File: plugins/services/src/js/events/SDKEndpointActions.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+          },
+          xhr
+        });
+      }
+    });
+  },
+
+  fetchEndpoint(serviceId, endpointName) {
+    const url = `/service/${serviceId}/v1/endpoints/${endpointName}`;
+    const fetch = new XMLHttpRequest();
```
[the remark this answers is not in the document]
Comment (bstavroulakis) [reply plugins/services/src/js/events/SDKEndpointActions.js]: Renamed to request

File: plugins/services/src/js/events/SDKEndpointActions.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    fetch.open("GET", url);
+    fetch.onreadystatechange = function() {
+      if (fetch.readyState === 4) {
```
[the remark this answers is not in the document]
Comment (bstavroulakis) [reply plugins/services/src/js/events/SDKEndpointActions.js]: Changed

File: plugins/services/src/js/events/SDKEndpointActions.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        if (fetch.status === 200) {
+          const contentType = fetch.getResponseHeader("Content-Type");
+
+          AppDispatcher.handleServerAction({
+            type: REQUEST_SDK_ENDPOINT_SUCCESS,
+            data: {
+              serviceId,
+              endpointData: contentType.includes("json")
+                ? JSON.parse(fetch.response)
```
[the remark this answers is not in the document]
Comment (bstavroulakis) [reply plugins/services/src/js/events/SDKEndpointActions.js]: Yes, would be better probably to keep raw data and parse this in the store. But there are the only 2 cases for now. Maybe we can move over the logic if this changes in the future.

File: plugins/services/src/js/events/SDKEndpointActions.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+const SDKEndpointsActions = {
+  fetchEndpoints(serviceId) {
+    const url = `/service/${serviceId}/v1/endpoints`;
+    RequestUtil.json({
+      url,
+      method: "GET",
+      success(xhr) {
+        AppDispatcher.handleServerAction({
+          type: REQUEST_SDK_ENDPOINTS_SUCCESS,
+          data: { serviceId, endpoints: xhr }
```
Comment (MatApple) [inline plugins/services/src/js/events/SDKEndpointActions.js]: Do we need to pass the xhr Object as part of the dispatch payload? Usually we only need to do that on the error case. The payload I'd expect to see here for endpoints would be more like `endpoints: RequestUtil.parseResponseBody(xhr)`

Comment (bstavroulakis) [reply plugins/services/src/js/events/SDKEndpointActions.js]: The parseResponseBody was created for responses that have a xhr.responseJSON obj or a xhr.responseText obj. In this case we return as a body the array of the endpoints so the parseResponseBody in the RequestUtil won't work for us and we can return the data immediately. 

- From the RequestUtil.parseResponseBody(xhr)
```
var responseJSON = xhr.responseJSON;
    var responseText = xhr.responseText;

    if (responseJSON) {
      return responseJSON;
    }
```

Comment (bstavroulakis) [reply plugins/services/src/js/events/SDKEndpointActions.js]: Renamed to "endpoints"

File: plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        expect(this.configuration.url).toEqual(
+          `${Config.rootUrl}/service/${serviceId}/v1/endpoints`
+        );
+      });
+
+      it("uses GET for the request method", function() {
+        expect(this.configuration.method).toEqual("GET");
+      });
+
+      it("dispatches the correct action when successful", function() {
```
Comment (MatApple) [inline plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: Please make sure to test that all expected properties are emitted on both success and error - not just `type`. Sorry I missed this on the first review 😞

Comment (bstavroulakis) [reply plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: Yes, listening now to all properties.

File: plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
+const RequestUtil = require("mesosphere-shared-reactjs").RequestUtil;
```
Comment (nLight) [inline plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: we actually can use `import` in tests as well

File: plugins/services/src/js/events/SDKEndpointActions.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      }
+    });
+  },
+
+  fetchEndpoint(serviceId, endpointName) {
+    const url = `/service/${serviceId}/v1/endpoints/${endpointName}`;
+    const request = new XMLHttpRequest();
+
+    request.open("GET", url);
+    request.onreadystatechange = function() {
```
[the remark this answers is not in the document]
Comment (bstavroulakis) [reply plugins/services/src/js/events/SDKEndpointActions.js]: Yes, good idea to be more explicit

File: plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js (hunk 7 above)
Comment (bstavroulakis) [reply plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: Used the import, it is unfortunate that most of the tests use the require.

Comment (orlandohohmeier) [reply plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: Yeah, we should update all of the test till then let's please be consistent with at least the resp. package/plugin.

Comment (bstavroulakis) [reply plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: changed back

File: plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+    it("call open with GET request method and correct url", function() {
+      const url = `/service/${serviceId}/v1/endpoints/${endpointName}`;
+
+      SDKEndpointsActions.fetchEndpoint(serviceId, endpointName);
+
+      expect(mockXhr.open).toHaveBeenCalledWith("GET", url);
+    });
+
+    it("dispatches the correct action when successful", function() {
```
Comment (ahoskins) [inline plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: Add another test case for properly handling a response that is *not* JSON?

Comment (bstavroulakis) [reply plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: Included a test case for text.

File: plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+            endpointName: "arangodb",
+            serviceId: "foo"
+          }
+        });
+      });
+
+      mockXhr.success();
+    });
+
+    it.only(
```
Comment (MatApple) [inline plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: Please remove the `.only` ;)

Comment (bstavroulakis) [reply plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: Ooops, removed.

File: plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js (hunk 7 above)
Comment (nLight) [reply plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: We could just do one by one, esp. for all new files.

Comment (orlandohohmeier) [reply plugins/services/src/js/events/__tests__/SDKEndpointsActions-test.js]: It's super easy to change them all at once. I'll file a JIRA and schedule for next sprint.

**Pull request merged**